zipline/algorithm.py

```diff
@@ -632,42 +632,96 @@ def run(self, data=None, overwrite_sim_params=True):
     def _write_and_map_id_index_to_sids(self, identifiers, as_of_date):
         # Build new Assets for identifiers that can't be resolved as
         # sids/Assets
-        identifiers_to_build = set()
-        next_sid = max(self.asset_finder.sids or (0,)) + 1
+        def is_unknown(asset_or_sid):
+            sid = op.index(asset_or_sid)
+            return self.asset_finder.retrieve_asset(
```

Comment:
My understanding is that after the bundle PR, the frame/panel case will get intercepted much earlier so that we no longer need to mutate an existing AssetFinder in place, which renders this obsolete.  The fix being made here is for a specific pattern that arises only on Q research where you have asset objects loaded from an external resource, but you don't have the full asset finder available.  I think the right fix for that is to have a proper `run_algorithm` function that hides the details of resource construction from the user.